src/app/CommandSender.cpp

```diff
@@ -62,24 +62,49 @@ CHIP_ERROR CommandSender::AllocateBuffer()
 
 CHIP_ERROR CommandSender::SendCommandRequest(const SessionHandle & session, System::Clock::Timeout timeout)
 {
-    VerifyOrReturnError(mState == State::AddedCommand, CHIP_ERROR_INCORRECT_STATE);
+    CHIP_ERROR err      = CHIP_NO_ERROR;
+    bool isGroupCommand = false;
+    VerifyOrExit(mState == State::AddedCommand, err = CHIP_ERROR_INCORRECT_STATE);
 
-    ReturnErrorOnFailure(Finalize(mPendingInvokeData));
+    err = Finalize(mPendingInvokeData);
+    SuccessOrExit(err);
 
     // Create a new exchange context.
     mpExchangeCtx = mpExchangeMgr->NewContext(session, this);
-    VerifyOrReturnError(mpExchangeCtx != nullptr, CHIP_ERROR_NO_MEMORY);
+    VerifyOrExit(mpExchangeCtx != nullptr, err = CHIP_ERROR_NO_MEMORY);
 
+    isGroupCommand = mpExchangeCtx->IsGroupExchangeContext();
     mpExchangeCtx->SetResponseTimeout(timeout);
 
     if (mTimedInvokeTimeoutMs.HasValue())
     {
-        ReturnErrorOnFailure(TimedRequest::Send(mpExchangeCtx, mTimedInvokeTimeoutMs.Value()));
+        VerifyOrExit(!isGroupCommand, err = CHIP_ERROR_INCORRECT_STATE);
+
+        err = TimedRequest::Send(mpExchangeCtx, mTimedInvokeTimeoutMs.Value());
+        SuccessOrExit(err);
         MoveToState(State::AwaitingTimedStatus);
-        return CHIP_NO_ERROR;
     }
+    else
+    {
+        err = SendInvokeRequest();
+    }
+    SuccessOrExit(err);
 
-    return SendInvokeRequest();
+exit:
+    if (mpCallback != nullptr)
+    {
+        if (err != CHIP_NO_ERROR)
+        {
+            mpCallback->OnError(this, StatusIB(Protocols::InteractionModel::Status::Failure), err);
+        }
+    }
+
+    if (isGroupCommand)
+    {
+        Close();
```

Comment:
This is, again, going to double-delete things in the error case, no?

What fails if the onDone is async?

If we _do_ call it sync, in success cases only, we need to:

1) Very clearly document that this can happen and under what conditions it happens.
2) Audit all consumers to make sure that they can handle this.

Reply:
- In the new implementation, Close() was moved before than exit. In case of error, it will not be called.
- Documentation was updated in the new implementation to fit with SendGroupCommandRequest